### DISCUSSION TITLE
September 2022 Project Update

### DISCUSSION THREAD

**Discussion by Kinostrome** [lexiconhq/lexicon#33] [Announcements]
As mentioned in ([this comment](https://github.com/lexiconhq/lexicon/issues/4#issuecomment-1198061742)) in #4, it would be beneficial to those interested in this project if we have a bit of an update on the current state of things

### Project History

First, let's start with a recap of everything.

As mentioned in the documentation on the [Background & Motivation](https://docs.lexicon.is/rationale/) page:

> In our consulting projects, we found that many of our clients were regularly asking for solutions that Discourse already provides out of the box.
> However, our clients wanted a seamless, native mobile experience, tailored to the brand that their users were already familiar with.

Lexicon was built primarily because of its usefulness to us with future client projects. (_This is important to keep in mind as you compare it to other Discourse related mobile apps & projects_.)

We felt that the open-source community may benefit from the work we had done as well, so we wanted to launch it as an open-source project.

The project officially launched in August of 2021. There seemed to be a lot of buzz and interest from a few of the channels that the announcement was published on.

There were a few aspects of the project which weren't included in the initial release, but intended to be included as follow-on work (as was outlined in the [documentation](https://docs.lexicon.is/discourse-features/#lexicon-mobile-app-features)):

- Push Notifications (mentioned as in-progress at the time)
- Discourse-specific features like posts, Discourse emojis, toggles, and better support for quotes
- The possibility of email deep-linking (not scoped)

Shortly after launch, priorities changed for us for a bit, in that all of our engineers contributing to this project were needed on other client projects. 

### 2022 Update

More recently, I finally found some availability to return to this project in my spare time. I had noticed that there was still interest from the community, and that people were having trouble getting past the quick start example.

#### Progress so far
<details><summary>Since I've revisited things, I've been focused on getting the code base back into a pristine state. <b>(Read more)</b></summary>

It should come as no surprise that—in the Javascript ecosystem—a project that was started in 2020 and launched in 2021 already has several dependencies that are quite out of date.

I've already updated all of the dependencies in the frontend: https://github.com/lexiconhq/lexicon/commit/97557a84ab1bbe7506b96f99121b8358b1d6437e. This was actually quite a lot of work, as it involved a lot of breaking API changes, new packages, and new type / linter errors that weren't previously identifiable.

This also included a fix for #1 (which was actually an internationalization issue).

I've also been responding to the huge backlog of dependabot warnings, and have that cut down to a much more reasonable number already.

After that, I realized that one of the dependabot PRs (upgrade axios from 0.19.2 to 0.21.2) actually created a bug, which was tracked in #5 and I recently pushed a fix for with https://github.com/lexiconhq/lexicon/commit/9422a1322ab04a076935e6757ad93982391c1560.

This bug caused any mutations (creating / editing topics and posts, etc.) to fail. Since that was a pretty significant regression, I made sure it was triaged and fixed ASAP. I have confirmed it is now working again as before.
</details>

It's also worth mentioning—I've made a few updates to make it easier to get started with pointing Lexicon at your Discourse site:

- The Prose GraphQL API now launches by default on the public interface `0.0.0.0` and port 80 (https://github.com/lexiconhq/lexicon/commit/97557a84ab1bbe7506b96f99121b8358b1d6437e). And the `.env.example` files now include this as a default as well.
- The README now contains an [additional small tutorial](https://github.com/lexiconhq/lexicon#example-specifying-a-custom-discourse-site-url) to walk you through the process step-by-step. 

### What comes next?

- Right now I'm in the middle of upgrading all dependencies and fixing any lingering regressions in the API. This is almost done.
- A few engineers have joined the project and are working hard on refactoring and tackling existing issues to provide a much better UX in app itself.
- Following that, we plan to move into providing push notifications and resume where left off. This will involve publishing a Discourse Plugin for Lexicon. It is not clear how much work is left, but we'll see what can be done.
- Everything else will be captured in a project roadmap, which will make it much easier to contextualize the status of the project and any features / issues you're interested in going forward.

**Comment by JuliaAgostino** [comment]
I'm glad to hear that this project hasn't been abandoned. I'll be closely following this repository now that development has resumed.

**Comment by Kinostrome** [comment]
> I'm glad to hear that this project hasn't been abandoned. I'll be closely following this repository now that development has resumed.

Could you tell us a bit more about how you're intending to use Lexicon? Have you deployed it anywhere yet?

**Comment by JuliaAgostino** [comment]
> > I'm glad to hear that this project hasn't been abandoned. I'll be closely following this repository now that development has resumed.
> 
> Could you tell us a bit more about how you're intending to use Lexicon? Have you deployed it anywhere yet?

No, I'm more of a Discourse enthusiast than a developer.
